ip_cidr_set.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"fmt"
+	"net"
+)
+
+// Fast lookup table for intersection of a single IP address within a collection of CIDR networks.
+//
+// Supports 4-byte (IPv4) and 16-byte (IPv6) networks.
+//
+// Provides O(1) best-case, O(log(n)) worst-case performance.
```

Comment:
Is this accurate? Or is it more accurately `O(k)` worst-case performance where `k` is the CIDR prefix (since that is what you are iterating through an array (k) and then having the contained `ipCIDRSet` doing the O(1) set operation).

And since  `k` is limited to 0-64, it's essentially a constant time operation anyway.

If you really wanted that operation to be pure speed O(1), you could replace the `[]ipNetMap` with a 33/65-length arrays (ipv4 v ipv6) and use the CIDR prefix bits count as the array index (your pure hashing function).

Probably overkill though since this list is essentially what users pass in on the command line as the `ip-whitelist` options? So it will likely be tiny.

Your worst case performance actually comes from the embedded set, since hashmaps have O(N) worst case performance on bad collisions in the hashring I believe.  But again, I'd like to see a user put in that many `ip-whitelist` option flags to make that collision scenario ever happen 😄

Reply:
So going by user behavior though, inputing a hand full of CIDR blocks to let through (or maybe at worst getting a long 100 block list from some cloud vendor) -- is all this complexity over optimizing at the expense of simpler code when all our use cases will be constant time regardless (N < 100)?

Reply:
> And since k is limited to 0-64, it's essentially a constant time operation anyway.

Not sure where the 64 number is coming from, maybe you mean 128?  But yeah, it depends on how you look at it.  It is yes constrained to < 128, but in the same sense that on a 64-bit system your classic binary tree in memory is also constrained to less than `2 ** 64` items and thus also has a constant worst case (< 64) and so could also be perceived as constant time.  At any rate, it's a lot better than `O(n)` and in the typical case will probably only require one to three lookups on the hash table.

> Your worst case performance actually comes from the embedded set, since hashmaps have O(N) worst case performance on bad collisions in the hashring I believe. But again, I'd like to see a user put in that many ip-whitelist option flags to make that collision scenario ever happen smile

Okay yeah that's true, I was operating under the assumption that hash maps are O(1) (which they generally are in reasonable cases).

> If you really wanted that operation to be pure speed O(1), you could replace the []ipNetMap with a 33/65-length arrays (ipv4 v ipv6) and use the CIDR prefix bits count as the array index (your pure hashing function).

Not sure where the 65 number is coming from?  Do you maybe mean 129?
Additionally 33/129 wouldn't happen anyways because in order to obtain 33/129 you nets you would need to include the `/0` netmask, thus including every address.

> So going by user behavior though, inputing a hand full of CIDR blocks to let through (or maybe at worst getting a long 100 block list from some cloud vendor) -- is all this complexity over optimizing at the expense of simpler code when all our use cases will be constant time regardless (N < 100)?

I might agree, except that in the oauth2-proxy case the IP whitelisting is checked on all requests, even those already authenticated.  If heavy traffic throughput goes through the proxy (even though it's mostly already authenticated), we'll still have to check around 100 (potentially) for every single request (even those using keep-alive).  I haven't produced benchmarks for it, so maybe it's useless, but I didn't want to introduce a burden into the code base.  We can remove if @JoelSpeed and @syscll don't think it's necessary.  It wasn't too hard to write so no-harm no-foul and the code can stick around in GitHub if we ever decide it's necessary.

Reply:
whoops, - definitely mean 128 whenever I said 64 😄

Reply:
> > If you really wanted that operation to be pure speed O(1), you could replace the []ipNetMap with a 33/65-length arrays (ipv4 v ipv6) and use the CIDR prefix bits count as the array index (your pure hashing function).
> 
> Not sure where the 65 number is coming from? Do you maybe mean 129?
> Additionally 33/129 wouldn't happen anyways because in order to obtain 33/129 you nets you would need to include the `/0` netmask, thus including every address.
> 

Might as well just include the unused `[0]` index and do 33 & 129 length arrays, then you wouldn't have to do `idx - 1` for every lookup (were you to even go this route, I just got geeked up about the possibility to use a raw array for hashing -- haven't had the need to do this in a while in real code besides coding interviews where the tricksters don't accept an O(N) worst case hash solution)

Reply:
> I might agree, except that in the oauth2-proxy case the IP whitelisting is checked on all requests, even those already authenticated. If heavy traffic throughput goes through the proxy (even though it's mostly already authenticated), we'll still have to check around 100 (potentially) for every single request (even those using keep-alive). I haven't produced benchmarks for it, so maybe it's useless, but I didn't want to introduce a burden into the code base. We can remove if @JoelSpeed and @syscll don't think it's necessary. It wasn't too hard to write so no-harm no-foul and the code can stick around in GitHub if we ever decide it's necessary.

Don't get me wrong, once I wrapped my head around it I thought it was a super clever optimization and I'm into it 👍

Reply:
Eh - never mind on the hashing array idea -- would be slower than your implementation, I think I have a better understanding now :+1: -- Good stuff

At whitelist lookup time, we want that array of potential prefix sizes to be small to iterate through (since we wouldn't know which prefix size would have the match for a given source IP that comes in, needs to check them all).

So in essence it checks in order of the unique `*/prefix` values provided by the user via options?

Reply:
Exactly, you got it!